File: src/ast/tree/function/parameter.ts
import type { Pass } from "../../../pass/pass";
import type { Dependent } from "../../node/expression";
import type { Lambda } from "../../node/lambda";
import { Node } from "../../node/node";
import type { Type } from "../../node/type";

export class Parameter extends Node<Type<Lambda> | Lambda> {
    constructor(
        public name: string,
        public type: Type<any> | null,
        public defaultValue: Dependent<any> | null,
        parent: Lambda | Type<Lambda> | null
    ) {
        super(parent);
    }

    public visit(pass: Pass): void {
        pass.visitParameter(this);
    }
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/function/Parameter.kt
package me.honkling.miniscript.parser.ast.function

import me.honkling.miniscript.diagnostic.Location
import me.honkling.miniscript.parser.ast.Node
import me.honkling.miniscript.parser.ast.Type
import me.honkling.miniscript.parser.ast.expression.Expression
import me.honkling.miniscript.pass.Pass

class Parameter(
    val name: String?,
    val type: Type<*>?,
    val defaultValue: Expression<*>?,
    val isVararg: Boolean = false,
    location: Location,
    parent: Node<*>?
) : Node<Node<*>?>(location, parent) {
    override fun accept(pass: Pass) {
        pass.visitParameter(this)
    }
}